### PR TITLE
WINC-815: Add support for Windows Server 2022 on GCP clusters

### PR DESCRIPTION
**Pull request by sebsoto**
This PR adds multiple commits around providing WMCO support for GCP clusters
* Commits to have GCP to work with our test suite:
   * [controllers] Add ignore label to filter Machines 
   * [e2e] Add GCP provider

*  Commits to add GCP functionality and docs to WMCO:
   * [secrets] Enable Administrator account if needed
   * [docs] Document GCP support

### CODE REVIEW

File: controllers/controllers.go [hunk 1]
```diff
@@ -162,10 +162,15 @@ func (r *instanceReconciler) reconcileKubeletClientCA(ctx context.Context, bundl
 }
 
 // GetAddress returns a non-ipv6 address that can be used to reach a Windows node. This can be either an ipv4
-// or dns address.
+// or dns address. DNS will be preferred, if available.
```
Comment (aravindhp) [inline controllers/controllers.go:165]: This change needs to be a separate PR to be on the safe side given it changes behavior across all clouds.

File: controllers/windowsmachine_controller.go [hunk 2]
```diff
@@ -202,7 +201,7 @@ func (r *WindowsMachineReconciler) isValidMachine(obj client.Object) bool {
 		return false
 	}
 
-	_, err := getInternalIPAddress(machine.Status.Addresses)
+	_, err := GetAddress(machine.Status.Addresses)
```
Comment (aravindhp) [inline controllers/windowsmachine_controller.go:209]: Where is `GetAddress` defined?

File: controllers/windowsmachine_controller.go [hunk 3]
```diff
@@ -53,6 +53,8 @@ const (
 	MachineOSLabel = "machine.openshift.io/os-id"
 	// WindowsMachineController is the name of this controller in logs and other outputs.
 	WindowsMachineController = "windowsmachine"
+	// IgnoreLabel is a label that will cause machines to be ignored by the Windows Machine controller
+	IgnoreLabel = "windowsmachineconfig.openshift.io/ignore"
```
Comment (aravindhp) [inline controllers/windowsmachine_controller.go:57]: You are couching this change as an `e2e` commit but you are actually changing the controller's behavior.

File: pkg/secrets/secrets.go [hunk 4]
```diff
@@ -52,6 +52,22 @@ func GenerateUserData(publicKey ssh.PublicKey) (*core.Secret, error) {
 		},
 		Data: map[string][]byte{
 			"userData": []byte(`<powershell>
+			function Get-RandomPassword {
+				Add-Type -AssemblyName 'System.Web'
+				return [System.Web.Security.Membership]::GeneratePassword(16, 2)
+			}
+
+			# Check if the capi user exists, this will be the case on Azure, and will be used instead of Administrator
```
Comment (aravindhp) [inline pkg/secrets/secrets.go:60]: As an aside I wonder if we can do this on Azure also? It is also making me wonder if we should just create a new `core` user instead like we wanted to in the past.

File: pkg/secrets/secrets.go [hunk 5]
```diff
@@ -52,6 +52,22 @@ func GenerateUserData(publicKey ssh.PublicKey) (*core.Secret, error) {
...
+			if((Get-LocalUser |Where-Object {$_.Name -eq "capi"})  -eq $null) {
+				# The capi user doesn't exist, ensure the Administrator account is enabled
+				$UserAccount = Get-LocalUser -Name "Administrator"
```
Comment (aravindhp) [inline pkg/secrets/secrets.go:63]: What happens if there is no Administrator account?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Thanks for working on this, @sebsoto. Looks good. Just some minor comments and splitting up the controller using DNS to another PR.

File: controllers/windowsmachine_controller.go (hunk 3 above)
Comment (sebsoto) [reply controllers/windowsmachine_controller.go:58]: Reworked the commit title to correct this

File: pkg/secrets/secrets.go (hunk 5 above)
Comment (sebsoto) [reply pkg/secrets/secrets.go:63]: It would fail and continue trying to execute the script if both the Administrator account and capi user are missing.
Changed it to only try to enable the Administrator account if it exists.

File: pkg/secrets/secrets.go (hunk 4 above)
Comment (sebsoto) [reply pkg/secrets/secrets.go:60]: We can re-open https://issues.redhat.com/browse/WINC-430
This could cause issues with people using custom images

File: controllers/windowsmachine_controller.go (hunk 3 above)
Comment (aravindhp) [reply controllers/windowsmachine_controller.go:58]: Please document this.

File: pkg/secrets/secrets.go (hunk 4 above)
Comment (aravindhp) [reply pkg/secrets/secrets.go:60]: >This could cause issues with people using custom images

Why would creating a core user cause issues?

File: pkg/secrets/secrets.go (hunk 5 above)
Comment (aravindhp) [reply pkg/secrets/secrets.go:63]: We should output an error if there is no capi and Administrator user. This hopefully we will somehow end up in the cloud console log.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: pkg/secrets/secrets.go (hunk 4 above)
Comment (sebsoto) [reply pkg/secrets/secrets.go:60]: I'm thinking of a case where the Administrator user has been set up with certain permissions in the image, or theres security software installed special casing the Administrator user. Its nothing we cant doc around.

File: pkg/secrets/secrets.go (hunk 5 above)
Comment (sebsoto) [reply pkg/secrets/secrets.go:63]: This line does actually write out an error if the user doesn't exist.
Its the reason I used `Where-Object` in the capi query, that one will not write out an error if capi isn't found.

Comment (sebsoto) [reply pkg/secrets/secrets.go:63]: I'll add a comment anyway

File: pkg/secrets/secrets.go [hunk 6]
```diff
@@ -52,6 +52,23 @@ func GenerateUserData(publicKey ssh.PublicKey) (*core.Secret, error) {
...
+				Add-Type -AssemblyName 'System.Web'
+				return [System.Web.Security.Membership]::GeneratePassword(16, 2)
+			}
+
+			# Check if the capi user exists, this will be the case on Azure, and will be used instead of Administrator
+			if((Get-LocalUser |Where-Object {$_.Name -eq "capi"})  -eq $null) {
+				# The capi user doesn't exist, ensure the Administrator account is enabled if it exists
+				# If neither users exist, an error will be written to the console, but the script will still continue
+				$UserAccount = Get-LocalUser -Name "Administrator"
+				if( ($UserAccount -ne $null) -and (!$UserAccount.Enabled) ) {
```
Comment (aravindhp) [inline pkg/secrets/secrets.go:65]: Nit extra spaces after `if(` and `)`.

File: controllers/windowsmachine_controller.go (hunk 3 above)
Comment (aravindhp) [reply controllers/windowsmachine_controller.go:58]: This has not been documented.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: controllers/windowsmachine_controller.go (hunk 3 above)
Comment (sebsoto) [reply controllers/windowsmachine_controller.go:58]: Missed this, done

**Review by saifshaikh48 — COMMENTED** [review]
Does the e2e [logic to list nodes](https://github.com/sebsoto/windows-machine-config-operator/blob/gcpCIReal/test/e2e/create_test.go#L487-L494) also need to be changed?

I think the ignore label introduction is a big enough change that a short design meeting to introduce it and discuss with the whole team would be helpful.

**Comment by sebsoto** [comment]
@saifshaikh48 Nope that logic does not need to be changed. It is not switching based on the presence of the Windows label

**Review by aravindhp — APPROVED** [review]
Nice work @sebsoto

File: README.md [hunk 7]
```diff
@@ -105,6 +105,9 @@ The following label has to be added to the Machine spec within the MachineSet sp
 
 Not having these labels will result in the Windows node not being marked as a worker.
 
+If the Machine spec has the label `windowsmachineconfig.openshift.io/ignore=true`, the Machine will be ignored by WMCO,
```
Comment (mansikulkarni96) [inline README.md:108]: should this information be moved to the hacking.md file? Wondering if this would be useful in the README if we don't expect users to use this other than for development and testing.

Comment (sebsoto) [reply README.md:108]: That's true, I was back and forth about this. I can move it there

File: test/e2e/validation_test.go [hunk 8]
```diff
@@ -310,7 +310,8 @@ func (tc *testContext) sshSetup() error {
 func (tc *testContext) runPowerShellSSHJob(name, command, ip string) (string, error) {
 	// Modify command to work when default shell is the newer Powershell version present on Windows Server 2022.
 	powershellDefaultCommand := command
-	if tc.CloudProvider.GetType() == config.VSpherePlatformType || tc.CloudProvider.GetType() == config.AzurePlatformType {
+	if tc.CloudProvider.GetType() == config.VSpherePlatformType ||
+		tc.CloudProvider.GetType() == config.GCPPlatformType || tc.CloudProvider.GetType() == config.AzurePlatformType {
 		powershellDefaultCommand = strings.ReplaceAll(command, "\\\"", "\"")
```
Comment (jrvaldes) [inline test/e2e/validation_test.go:315]: @sebsoto the cloud provider is not the driving factor here. This modification is required due to Windows Server 2022.

Consider using the existing `tc.windowsServerVersion` flag in the condition statement for a more accurate decision. See this [commit](https://github.com/openshift/windows-machine-config-operator/pull/1091/commits/c312195b164d86f351c3a5b5f98ffacd1d795bbf#diff-19a6ecdd2ef106d645523b5e6ddc22fd6ddfd1608477cd0b0fbd85113c8c245d).

For example:
```go
	if tc.windowsServerVersion != "2019" {
		powershellDefaultCommand = strings.ReplaceAll(command, "\\\"", "\"")
	}
```

File: docs/HACKING.md [hunk 9]
```diff
@@ -249,3 +249,8 @@ git submodule update --recursive
 # To update a specific submodule
 git submodule update --remote <path_to_submodule>
 ```
+
+## Preventing Windows Machines from being configured by WMCO
+
+If the Machine spec has the label `windowsmachineconfig.openshift.io/ignore=true`, the Machine will be ignored by WMCO,
+and will not be configured into a Windows node. This can be helpful when debugging userdata changes.
```
Comment (saifshaikh48) [inline docs/HACKING.md:256]: > the Machine will be ignored by WMCO and will not be configured into a Windows node

unless it is added as an entry to the windows-instances ConfigMap, right? Since this is hack readme, might be good to call this out.

File: pkg/secrets/secrets.go [hunk 10]
```diff
@@ -52,6 +52,23 @@ func GenerateUserData(publicKey ssh.PublicKey) (*core.Secret, error) {
 		},
 		Data: map[string][]byte{
 			"userData": []byte(`<powershell>
+			function Get-RandomPassword {
+				Add-Type -AssemblyName 'System.Web'
+				return [System.Web.Security.Membership]::GeneratePassword(16, 2)
+			}
+
+			# Check if the capi user exists, this will be the case on Azure, and will be used instead of Administrator
+			if((Get-LocalUser |Where-Object {$_.Name -eq "capi"})  -eq $null) {
```
Comment (saifshaikh48) [inline pkg/secrets/secrets.go:61]: nits: please put a space after the pipe and remove the extra space before `-eq` (there's 2)

**Review by saifshaikh48 — COMMENTED** [review]
Good stuff, just a few minor comments

File: docs/machineset-gcp.md [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+```shell script
+ oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
+```
+_\<region\>_ should be replaced with a region, such as `us-central1`.
+
+_\<zone\>_ should be replaced with a zone within the chosen region, such as `us-central1-a`.
+
+_\<zone_suffix\>_ should be replaced with the suffix of the chosen zone, such as `a`.
+
+_\<project_id\>_ should be replaced with the GCP project this cluster was created under.
```
Comment (jrvaldes) [inline docs/machineset-gcp.md:18]: Is there a `gcloud`  command to fetch this info? `gcloud config get-value project` maybe?

File: test/e2e/validation_test.go (hunk 8 above)
Comment (sebsoto) [reply test/e2e/validation_test.go:315]: @jrvaldes thats a good point, but I don't think that method is usable until the AWS 2022 PR is merged. If I try and do that now, the AWS 2022 jobs will fail

Comment (sebsoto) [reply test/e2e/validation_test.go:315]: What I mean is the jobs that are set for use with 2022 will fail due to the wrong powershell parsing being used for them

File: docs/HACKING.md (hunk 9 above)
Comment (sebsoto) [reply docs/HACKING.md:256]: I'm avoiding calling that out specifically because I dont want anyone to assume thats okay to do.
I'll change the language to specifically call out the Machine controller will ignore the Machine though.

File: docs/machineset-gcp.md (hunk 11 above)
Comment (sebsoto) [reply docs/machineset-gcp.md:18]: This could give the wrong value if the default project they have set up on their shell isn't the one that the cluster exists on

File: test/e2e/validation_test.go (hunk 8 above)
Comment (jrvaldes) [reply test/e2e/validation_test.go:315]: The environment variable introduced in https://github.com/openshift/release/pull/29791 and is ready to use.

File: docs/machineset-gcp.md (hunk 11 above)
Comment (jrvaldes) [reply docs/machineset-gcp.md:18]: I follow, is it safe to use it in the [machineset.sh](https://github.com/openshift/windows-machine-config-operator/blob/258b568146e2e7e2637aa0c28a0062bf34f6d20b/hack/machineset.sh#L189) script?

Comment (sebsoto) [reply docs/machineset-gcp.md:18]: i think its fine to use there as that script is meant for our use, not for customers

File: test/e2e/validation_test.go (hunk 8 above)
Comment (sebsoto) [reply test/e2e/validation_test.go:315]: Yes but aws-e2e-ccm-install and the aws upgrade job have that env var set to ""
this will cause the test suite to detect them as running server 2022 and use the wrong powershell format because of that

Comment (jrvaldes) [reply test/e2e/validation_test.go:315]: I don't think so. the Windows Server 2019 is [hard-coded](https://github.com/openshift/windows-machine-config-operator/blob/f619b40ec2175f58926547a2e2b94e5b1d8a22c0/test/e2e/providers/aws/aws.go#L128).

My point is that we should stop using `tc.CloudProvider.GetType()` as criteria to tweak the PS commands. It was valid with AWS+WS2022 and is now a reality with GCP+WS2022.

Comment (sebsoto) [reply test/e2e/validation_test.go:315]: we cannot do `powershellDefaultCommand = strings.ReplaceAll(command, "\\\"", "\"")` on VMs running Windows server 2019. The AWS upgrade and ccm jobs are currently running 2019, but checking that environment variable will tell us that they are running 2022.

**Comment by sebsoto** [comment]
Azure CI failing due to node logs test issues.
Opened https://github.com/openshift/windows-machine-config-operator/pull/1214 to address this
